src/mito2/src/memtable/bulk/part.rs

```diff
@@ -0,0 +1,666 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//!  Bulk part encoder/decoder.
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use api::v1::Mutation;
+use datafusion::arrow::array::{TimestampNanosecondArray, UInt64Builder};
+use datatypes::arrow;
+use datatypes::arrow::array::{
+    Array, ArrayData, ArrayRef, BinaryBuilder, DictionaryArray, RecordBatch,
+    TimestampMicrosecondArray, TimestampMillisecondArray, TimestampSecondArray, UInt16Array,
+    UInt16BufferBuilder, UInt32Array, UInt64Array, UInt8Array, UInt8Builder,
+};
+use datatypes::arrow::compute::TakeOptions;
+use datatypes::arrow::datatypes::{DataType as ArrowDataType, SchemaRef, UInt16Type};
+use datatypes::arrow_array::BinaryArray;
+use datatypes::data_type::DataType;
+use datatypes::prelude::{MutableVector, ScalarVectorBuilder, Vector};
+use datatypes::types::TimestampType;
+use parquet::arrow::ArrowWriter;
+use parquet::data_type::AsBytes;
+use snafu::ResultExt;
+use store_api::metadata::RegionMetadataRef;
+
+use crate::error::{ComputeArrowSnafu, EncodeMemtableSnafu, NewRecordBatchSnafu, Result};
+use crate::memtable::key_values::KeyValuesRef;
+use crate::read::Batch;
+use crate::row_converter::{McmpRowCodec, RowCodec};
+use crate::sst::to_sst_arrow_schema;
+
+#[derive(Debug)]
+pub struct BulkPart {
+    data: Vec<u8>,
+    metadata: BulkPartMeta,
+}
+
+impl BulkPart {
+    pub fn new(data: Vec<u8>, metadata: BulkPartMeta) -> Self {
+        Self { data, metadata }
+    }
+
+    pub(crate) fn metadata(&self) -> &BulkPartMeta {
+        &self.metadata
+    }
+}
+
+#[derive(Debug)]
+pub struct BulkPartMeta {
+    pub num_rows: usize,
+    pub size: usize,
+    pub max_timestamp: i64,
+    pub min_timestamp: i64,
+}
+
+impl Default for BulkPartMeta {
+    fn default() -> Self {
+        Self {
+            num_rows: 0,
+            size: 0,
+            max_timestamp: i64::MIN,
+            min_timestamp: i64::MAX,
+        }
+    }
+}
+
+pub struct BulkPartEncoder {
+    metadata: RegionMetadataRef,
+    arrow_schema: SchemaRef,
+    pk_encoder: McmpRowCodec,
+    dedup: bool,
```

Comment:
> non-blocking: Maybe we need to find a better way to reuse the dedup logic.

The dedup code here is slightly different from existing ones in that it compares tuples by tuples rather than encode columns to comparable rows.